aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,173 @@
+
+In this sample application, I'm using Entity Framework Core's InMemoryDatabase support, so I can't just connect to it from my test project. Instead, I expose an ``InitializeDatabase`` method from the app's ``Startup`` class, which I call when the app starts up if it's in the ``Development`` environment. My integration tests automatically benefit from this as long as they set the environment to ``Development``. I don't have to worry about resetting the database, since the InMemoryDatabase is reset each time the app restarts.
+
+The ``Startup`` class:
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/Startup.cs
+  :language: c#
+  :emphasize-lines: 18-21,40-41,49,57
+
+You'll see the ``GetTestSession`` method used frequently in the integration tests below.
+
+Accessing Views
+^^^^^^^^^^^^^^^
+Each integration test class configures the ``TestServer`` that will run the ASP.NET Core app. By default, ``TestServer`` hosts the web app in the folder where it's running - in this case, the test project folder. Thus, when you attempt to test controller actions that return ``ViewResult``, you may see this error:
+
+.. code-block:: c#
+
+	The view 'Index' was not found. The following locations were searched:
+	(list of locations)
+
+To correct this issue, you need to configure the server to use the ``ApplicationBasePath`` and ``ApplicationName`` of the web project. This is done in the call to ``UseServices`` in the integration test class shown:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/IntegrationTests/HomeControllerIndex.cs
+  :language: c#
+  :emphasize-lines: 20,22-32,37-38
+
+In the test above, the ``responseString`` gets the actual rendered HTML from the View, which can be inspected to confirm it contains expected results.
+
+API Methods
+^^^^^^^^^^^
+If your app exposes web APIs, it's a good idea to have automated tests confirm they execute as expected. The built-in ``TestServer`` makes it easy to test web APIs. If your API methods are using model binding, you should always check ``ModelState.IsValid``, and integration tests are the right place to confirm that your model validation is working properly. 
+
+The following set of tests target the ``Create`` method in the :ref:`IdeasController <ideas-controller>` class shown above:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/IntegrationTests/ApiIdeasControllerCreatePost.cs
+  :language: c#
+  :lines: 40-93
+  :dedent: 8
+  :emphasize-lines: 1-2,9-10,17-18,25-26,33-34,42-43,51
+
+Unlike integration tests of actions that returns HTML views, web API methods that return results can usually be cast to strongly typed objects, as the last test above shows. In this case, the test casts the result to a ``BrainstormSession`` instance, and confirms that the idea was correctly added to its collection of ideas.
```

Comment:
> web API methods that return results can usually be cast to strongly typed objects, as the last test above shows. In this case, the test casts the result to a ``BrainstormSession`` instance

Web API methods that return results can usually be deserialized as strongly typed objects, as the last test above shows. In this case, the test deserializes the result to a ``BrainstormSession`` instance